### PR TITLE
Atomic upsertWithWhere

### PR DESCRIPTION
**Pull request by mrbatista**
Implement atomic upsertWithWhere method.

The out-of-the-box implementation catch error if multiple instances are founds.
With atomic implementation one test present in the juggler datasource will be disabled.
I added a new option to specify if connector has an atomic implementation. See https://github.com/strongloop/loopback-datasource-juggler/pull/1864.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-connector-mongodb) 👈

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](https://loopback.io/doc/en/contrib/style-guide-es6.html)
- [x] Commit messages are following our [guidelines](https://loopback.io/doc/en/contrib/git-commit-messages.html)

### CODE REVIEW

**Comment by agnes512** [comment]
@mrbatista Thank you for the PR. Could you add some tests for the changes?

**Comment by dhmlau** [comment]
We just switch the contribution method from CLA to DCO, making your contribution easier in the future. Please sign the commits with DCO by amending your commit messages with `-s` flag and push the changes again. If you're using the command line, you can:

```
git commit --amend -s
git push --force-with-lease
```

Please refer to [this docs page](https://loopback.io/doc/en/contrib/code-contrib.html#developer-certificate-of-origin-dco) for details. Thanks! 

For more questions, please join our [Slack workspace - #loopback-contributors](https://loopbackio.slack.com/join/shared_invite/zt-8lbow73r-SKAKz61Vdao~_rGf91pcsw#/).

**Comment by jamesjjk** [comment]
@mrbatista @fabien I don't understand why you would want to use a transactional operation? Maybe I'm missing something. Atomic upsert and document return is achieved with findAndModify operation with the "new" and "update" flags. transactional seems overkill.

**Comment by mrbatista** [comment]
@jamesjjk If the connector does not implement the upsertWithWhere method, the DAO use a non atomic implementation. [See](https://github.com/strongloop/loopback-datasource-juggler/blob/261dd1c865e8b197882a161b909b90b0e6c3c2a3/lib/dao.js#L848) the relevant code.
I need the atomic implementation for concurrent requests.

**Comment by jamesjjk** [comment]
> @jamesjjk If the connector does not implement the upsertWithWhere method, the DAO use a non atomic implementation. [See](https://github.com/strongloop/loopback-datasource-juggler/blob/261dd1c865e8b197882a161b909b90b0e6c3c2a3/lib/dao.js#L848) the relevant code.
> I need the atomic implementation for concurrent requests.

@mrbatista Atomic I fully understand but why transactional? I too need an atomic version which is implemented by MongoDB native findAndModify with the appropriate parameters. Your ticket mentions transactional, which is not the same as an atomic opetation. Would it not be sufficient to implement findAndModify (with new and upsert parameters) in the connector to parse any of the DOA's upsertWithWhere operations?

**Comment by mrbatista** [comment]
@jamesjjk I changed the title to respect the implementation of this PR.
For the implementation the DAO define the methods that each connector (oracle, mysql, ecc..) can implement.
If use `findAndModfy` (with specific options set) or `updateMany` (like in this PR) is not relevant.

**Comment by jamesjjk** [comment]
@mrbatista Understood, are you updating the respective connector? Which connector are you using if I may ask?

**Comment by mrbatista** [comment]
> @mrbatista capito, stai aggiornando il rispettivo connettore? Quale connettore stai usando se posso chiedere?

I'm currently using my fork which implements atomic upsertWithWhere.
Today add more test to permit review by mainteners of this repo. :smiley:

**Comment by mrbatista** [comment]
@agnes512 I have update the PR and first comment. Please review.
No test needed because the functionality of upsertWithWhere method is tested in [loopback-datasource-juggler](https://github.com/strongloop/loopback-datasource-juggler).
See relevant test:
1. [normalize-undefined test](https://github.com/strongloop/loopback-datasource-juggler/blob/251ced811b80177f47efb6a3270ed6d279263c46/test/normalize-undefined.test.js)
2. [crud-with-options test](https://github.com/strongloop/loopback-datasource-juggler/blob/251ced811b80177f47efb6a3270ed6d279263c46/test/crud-with-options.test.js)
3. [manipulation test](https://github.com/strongloop/loopback-datasource-juggler/blob/2f867576b66fe535aa1f82bceaaa680ecd4eec6c/test/manipulation.test.js)
4. [persistence-hooks suite](https://github.com/strongloop/loopback-datasource-juggler/blob/251ced811b80177f47efb6a3270ed6d279263c46/test/persistence-hooks.suite.js)

**Comment by jamesjjk** [comment]
> > @mrbatista Understood, are you updating the respective connector? Which connector are you using if I may ask?
> 
> I'm currently using my fork which implements atomic upsertWithWhere.
> Today add more test to permit review by mainteners of this repo. 😃

Willing to share your fork of the connector? 👍

**Comment by mrbatista** [comment]
@jamesjjk Is the same of this PR -> https://github.com/mrbatista/loopback-connector-mongodb/tree/feat/upsert-with-where

**Comment by mrbatista** [comment]
@strongloop/loopback-maintainers ping

**Comment by bajtos** [comment]
Sorry for the delay, I'll try to take a look next week.

@jannyHou @hacksparrow could you PTAL too?

**Review by jannyHou — COMMENTED** [review]
@mrbatista Great optimization! For https://github.com/strongloop/loopback-datasource-juggler/pull/1864 the scenario is still valid I think, just we need to distinguish the result based on the `atomic` flag. Could you add a test to verify when:
- multiple instances found
- `atomicUpsertWithWhere` is true

`upsertWithWhere` successfully update only one record?
Thanks.

The method itself LGTM 👍

**Comment by mrbatista** [comment]
@jannyHou I added a test in juggler PR when atomicUpsertWithWhere is true. Please take a look.
The atomic implementation update the first instance that matching the criteria.
In the juggler PR I have added a comment that explains that the logic depends on the specific implementation of the connector.

**Comment by jannyHou** [comment]
@slnode test please

**Comment by jannyHou** [comment]
Thanks @mrbatista , I merged and published the juggler test, but it fails due to

```
1) juggler-v3

       manipulation

         upsertWithWhere

           fails the upsertWithWhere operation when multiple instances are retrieved based on the filter criteria:

      Uncaught AssertionError: expected 'Performing an update on the path \'_id\' would modify the immutable field \'_id\'' to be 'There are multiple instances found.Upsert Operation will not be performed!'

      + expected - actual

      -Performing an update on the path '_id' would modify the immutable field '_id'

      +There are multiple instances found.Upsert Operation will not be performed!
```

juggler-v3 complains the same problem. Could you take a look?

**Comment by mrbatista** [comment]
@jannyHou You have release only juggler 4.x. Please backport the change of [this](https://github.com/strongloop/loopback-datasource-juggler/commit/76acf333fdc2cbcb482af999e7503bbc63b80ad6) commit in 3.x juggler branch.
Ping me when release a new 3.x version.

**Comment by mrbatista** [comment]
@jannyHou On my local environment only the juggler 3.x fail.

**Comment by jannyHou** [comment]
@mrbatista Thanks, I see created PR https://github.com/strongloop/loopback-datasource-juggler/pull/1871 to back port it.

**Comment by mrbatista** [comment]
@jannyHou Any update on this PR?

**Comment by jannyHou** [comment]
Hi @mrbatista sorry for late update, I am making time to address the feedback in https://github.com/strongloop/loopback-datasource-juggler/pull/1871, after that, this PR should be good to land.

**Comment by jannyHou** [comment]
@slnode test please

**Comment by bajtos** [comment]
I wanted to re-run Travis CI build but it looks like the old Travis at https://travis-ci.org does not allow that (or I don't have permissions to do so). I think we will need to force-push to the feature branch to trigger a new build, can you @jannyHou PTAL?

We need a new run to pick the recently released versions of juggler v3 and v4 that should fix the failure of the test _it fails the upsertWithWhere operation when multiple instances are retrieved based on the filter criteria_

**Comment by jannyHou** [comment]
@bajtos thank you for looking into the error, triggered a test.

**Comment by jannyHou** [comment]
It failed because I didn't release the change in juggler v3. Just released 3.36.1, let me rerun it

**Comment by mrbatista** [comment]
@jannyHou green CI 🎉🥳

**Review by jannyHou — APPROVED** [review]
👍 Cheers!

**Review by agnes512 — APPROVED** [review]

**Review by bajtos — APPROVED** [review]
👏🏻

**Pull request merged**
